core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java

```diff
@@ -681,41 +686,64 @@ public TaskOutputs calculateTaskOutputs(TaskInputs taskInputs, ClusterState prev
                     e);
             }
             warnAboutSlowTaskIfNeeded(executionTime, taskInputs.summary);
-            batchResult = BatchResult.builder()
+            clusterTaskResult = ClusterTaskResult.builder()
                 .failures(taskInputs.updateTasks.stream().map(updateTask -> updateTask.task)::iterator, e)
                 .build(previousClusterState);
         }
 
-        assert batchResult.executionResults != null;
-        assert batchResult.executionResults.size() == taskInputs.updateTasks.size()
+        assert clusterTaskResult.executionResults != null;
+        assert clusterTaskResult.executionResults.size() == taskInputs.updateTasks.size()
                 : String.format(Locale.ROOT, "expected [%d] task result%s but was [%d]", taskInputs.updateTasks.size(),
-                taskInputs.updateTasks.size() == 1 ? "" : "s", batchResult.executionResults.size());
+                taskInputs.updateTasks.size() == 1 ? "" : "s", clusterTaskResult.executionResults.size());
         boolean assertsEnabled = false;
         assert (assertsEnabled = true);
         if (assertsEnabled) {
             for (UpdateTask updateTask : taskInputs.updateTasks) {
-                assert batchResult.executionResults.containsKey(updateTask.task) :
+                assert clusterTaskResult.executionResults.containsKey(updateTask.task) :
                     "missing task result for " + updateTask;
             }
         }
 
-        return batchResult;
+        return clusterTaskResult;
     }
 
-    private ClusterState patchVersions(ClusterState previousClusterState, ClusterState newClusterState) {
-        if (previousClusterState != newClusterState) {
-            if (newClusterState.nodes().isLocalNodeElectedMaster()) {
-                // only the master controls the version numbers
-                Builder builder = ClusterState.builder(newClusterState).incrementVersion();
-                if (previousClusterState.routingTable() != newClusterState.routingTable()) {
-                    builder.routingTable(RoutingTable.builder(newClusterState.routingTable())
-                            .version(newClusterState.routingTable().version() + 1).build());
-                }
-                if (previousClusterState.metaData() != newClusterState.metaData()) {
-                    builder.metaData(MetaData.builder(newClusterState.metaData()).version(newClusterState.metaData().version() + 1));
-                }
-                newClusterState = builder.build();
+    private ClusterState patchBlocksAndVersions(ClusterState previousClusterState, ClusterTaskResult<Object> executionResult) {
+        ClusterState newClusterState = executionResult.resultingState;
+        if (newClusterState == null) {
```

Comment:
I've changed the LocalClusterUpdateTask to put the previous state when the cluster state is null. This means that we don't have to handle this oddity in ClusterService anymore.